crates/script/src/build.rs

```diff
@@ -0,0 +1,231 @@
+    pub predeploy_libraries: Vec<Bytes>,
+}
+
+impl LinkedBuildData {
+    pub fn new(link_output: LinkOutput, build_data: BuildData) -> Result<Self> {
+        let highlevel_known_contracts = build_data
+            .linker
+            .get_linked_artifacts(&link_output.libraries)?
+            .iter()
+            .filter_map(|(id, contract)| {
+                ContractBytecodeSome::try_from(ContractBytecode::from(contract.clone()))
+                    .ok()
+                    .map(|tc| (id.clone(), tc))
+            })
+            .filter(|(_, tc)| tc.bytecode.object.is_non_empty_bytecode())
+            .collect();
+
+        Ok(Self {
+            build_data,
+            highlevel_known_contracts,
+            libraries: link_output.libraries,
+            predeploy_libraries: link_output.libs_to_deploy,
+        })
+    }
+
+    /// Flattens the contracts into  (`id` -> (`JsonAbi`, `Vec<u8>`)) pairs
+    pub fn get_flattened_contracts(&self, deployed_code: bool) -> ContractsByArtifact {
+        ContractsByArtifact(
+            self.highlevel_known_contracts
+                .iter()
+                .filter_map(|(id, c)| {
+                    let bytecode = if deployed_code {
+                        c.deployed_bytecode.bytes()
+                    } else {
+                        c.bytecode.bytes()
+                    };
+                    bytecode.cloned().map(|code| (id.clone(), (c.abi.clone(), code.into())))
+                })
+                .collect(),
+        )
+    }
+
+    /// Fetches target bytecode from linked contracts.
+    pub fn get_target_contract(&self) -> Result<ContractBytecodeSome> {
+        self.highlevel_known_contracts
+            .get(&self.build_data.target)
+            .cloned()
+            .ok_or_eyre("target not found in linked artifacts")
+    }
+}
+
+impl PreprocessedState {
+    /// Parses user input and compiles the contracts depending on script target.
+    /// After compilation, finds exact [ArtifactId] of the target contract.
+    pub fn compile(self) -> Result<CompiledState> {
+        let Self { args, script_config, script_wallets } = self;
+        let project = script_config.config.project()?;
+        let filters = args.skip.clone().unwrap_or_default();
+
+        let mut target_name = args.target_contract.clone();
+
+        // If we've received correct path, use it as target_path
+        // Otherwise, parse input as <path>:<name> and use the path from the contract info, if
+        // present.
+        let target_path = if let Ok(path) = dunce::canonicalize(&args.path) {
+            Ok::<_, eyre::Report>(Some(path))
+        } else {
+            let contract = ContractInfo::from_str(&args.path)?;
+            target_name = Some(contract.name.clone());
+            if let Some(path) = contract.path {
+                Ok(Some(dunce::canonicalize(path)?))
+            } else {
+                Ok(None)
+            }
+        }?;
```

Comment:
is the additional Ok wrapper required here?
I dont think we have an err, hence the `Ok<_, eyre>`